File: app/src/main/java/tag_them/metoothanks/items/Text.java
package tag_them.metoothanks.items;

import android.graphics.Canvas;

import tag_them.metoothanks.CanvasView;
import tag_them.metoothanks.R;
import tag_them.metoothanks.items.subitems.TextString;

import static tag_them.metoothanks.Utils.calculateTextHeight;
import static tag_them.metoothanks.Utils.calculateTextWidth;
import static tag_them.metoothanks.items.subitems.TextString.DEFAULT_TEXT_SIZE;

public class Text extends Item {
	private TextString textString;
	
	private final int WIDTH_FIX_INTERVAL = 10;
	
	public Text(String text, final int canvasWidth, CanvasView hostView) {
		super(hostView, calculateTextWidth(text, DEFAULT_TEXT_SIZE), calculateTextHeight(text, DEFAULT_TEXT_SIZE));
		textString = new TextString(text);
		
		// makes sure the actual text fits the item's width
		new Thread(() -> {
			while (true) {
				try {
					Thread.sleep(WIDTH_FIX_INTERVAL);
					
					int width = Math.min(getWidth(), canvasWidth);
					textString.fitToWidth(width);
					right = left + width;
					
					int textHeight = calculateTextHeight(text, textString.getFontSize());
					if (getHeight() < textHeight)
						bottom = top + textHeight;
				} catch (Exception e) {
					e.printStackTrace();
				}
			}
		}).start();
	}
	
	
	@Override
	public void draw(Canvas canvas) {
		textString.draw(left, top, right, canvas);
		drawFrame(canvas);
	}
	
	@Override
	public CanvasView.ClickListener getClickListener() {
		return item -> {
			switch (item.getItemId()) {
				case R.id.action_text_align_left:
					textString.setAlignment(TextString.ALIGNMENT_LEFT);
					
					
					hostView.postInvalidate();
					return true;
				
				case R.id.action_text_align_right:
					textString.setAlignment(TextString.ALIGNMENT_RIGHT);
					
					hostView.postInvalidate();
					return true;
				
				case R.id.action_item_edit:
					textString.edit(hostView.getContext());
					
					hostView.postInvalidate();
					return true;
				
				case R.id.action_text_shrink:
					textString.decreaseFontSize();
					
					hostView.postInvalidate();
					return true;
				
				case R.id.action_text_enlarge:
					textString.increaseFontSize();
					
					int textWidth = textString.getWidth();
					if (textWidth > getWidth())
						right = left + textWidth;
					
					hostView.postInvalidate();
					return true;
			}
			return false;
		};
	}
	
	@Override
	public int getMenuID() {
		return R.menu.item_text_menu;
	}
	
	@Override
	public String toString() {
		return textString.toString();
	}
}

File: app/src/main/java/org/tag_them/metoothanks/items/Text.kt
package org.tag_them.metoothanks.items

import android.graphics.Canvas
import android.graphics.Point
import android.text.TextPaint
import android.view.MenuItem
import org.tag_them.metoothanks.*


val DEFAULT_TEXT_SIZE = 120f
val MIN_TEXT_SIZE = 40
val TEXT_SIZE_CHANGE_STEP = 20
val WIDTH_FIX_INTERVAL = 10L

class Text(text: String, canvasWidth: Int, hostView: CanvasView) :
        Item(
                canvas = hostView,
                width = minOf(canvasWidth, calculateTextWidth(text, DEFAULT_TEXT_SIZE)),
                height = calculateTextHeight(text, applyFontHeightFix(DEFAULT_TEXT_SIZE))
        ) {

        var text: String = text
                private set(value) {
                        field = value
                        matchBoundariesWithText()
                }

        private val textPaint: TextPaint = TextPaint().apply {
                isAntiAlias = true
                textSize = DEFAULT_TEXT_SIZE
                color = 0xFF000000.toInt()
        }

        private enum class Alignment { LTR, RTL }

        private var alignment: Alignment = Alignment.LTR

        init {
                this.text = fitToWidth(canvasWidth, text, fontSize)
        }

        private var currentTextWidth = calculateTextWidth(text, fontSize)

        override fun handleMenuItemClick(item: MenuItem): Boolean {
                when (item.itemId) {
                        R.id.action_text_align_left  -> alignment = Alignment.LTR
                        R.id.action_text_align_right -> alignment = Alignment.RTL
                        R.id.action_item_edit        -> edit()
                        R.id.action_text_enlarge     -> increaseFontSize()
                        R.id.action_text_shrink      -> decreaseFontSize()
                        else                         -> return false
                }

                canvas.postInvalidate()

                return true
        }

        override val itemMenuID: Int = R.menu.item_text_menu

        override fun draw(canvas: Canvas) {
                val x = (if (alignment == Alignment.LTR) left else right).toFloat()
                val lines = text.split("\n")
                lines.indices.forEach {
                        canvas.drawText(lines[it], x, top + (it + 1) * applyFontHeightFix(fontSize), textPaint)
                }
        }

        override fun resize(pointers: PointPair, pointersGrip: PointPair,
                            pointersGripDistance: PointPair) {
                for (index in 0..1) {
                        if (pointersGrip[index].x in 0..(left + width / 2))
                                with(pointers[index].x - pointersGripDistance[index].x) {
                                        left = this
                                }
                        else
                                with(pointers[index].x + pointersGripDistance[index].x) {
                                        right = this
                                }

                        if (width < calculateTextWidth(text, fontSize)) {
                                text = fitToWidth(width, text, fontSize)
                                matchBoundariesWithText()
                        }
                }
        }

        fun edit() {
                canvas.context.openTextInputDialog(initialText = text) {
                        text = it
                        matchBoundariesWithText()
                }
        }

        private fun matchBoundariesWithText() {
                bottom = top + calculateTextHeight(text, applyFontHeightFix(fontSize))
                currentTextWidth = calculateTextWidth(text, fontSize)

                // comparing the current frame width to the new text's width
                if (width < currentTextWidth)
                        when (alignment) {
                                Alignment.LTR -> right = left + currentTextWidth
                                else          -> left = right - currentTextWidth
                        }
        }

        private fun changeFontSize(difference: Int) {
                textPaint.textSize += difference
                matchBoundariesWithText()
        }

        private fun increaseFontSize() {
                changeFontSize(TEXT_SIZE_CHANGE_STEP)
        }

        private fun decreaseFontSize() {
                if (fontSize > MIN_TEXT_SIZE + TEXT_SIZE_CHANGE_STEP)
                        changeFontSize(-TEXT_SIZE_CHANGE_STEP)
        }

        private val fontSize: Float
                get() = textPaint.textSize

        override fun toString(): String = text
}


/* helper functions */

/**
 * TextPaint.textSize is a bit larger than the actual
 * height of the font. This is a dirty hack around it.
 */
fun applyFontHeightFix(originalFontSize: Float): Float {
        val fixedFactor = 0.25f
        return originalFontSize * (1 - fixedFactor)
}

fun String.lastLine() = this.split("\n").last()

fun fitToWidth(width: Int, text: String, textSize: Float): String {
        tailrec fun fitToWidth1(width: Int, text: String, words: List<String>, textSize: Float): String =
                if (words.isEmpty()) text.trim()
                else {
                        val newWidth = calculateTextWidth(text.lastLine() + words.first(), textSize)
                        val newText = text + (if (newWidth > width) "\n" else " ") + words.first()
                        fitToWidth1(width, newText, words.rest(), textSize)
                }

        return fitToWidth1(width, "", text.split(" "), textSize)
}
